### PR TITLE
[Feature] Grading System - Participant Scores

### PR DESCRIPTION
**Pull request by ivohashamov**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, editor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all newly inserted strings into English and German.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).

### Motivation and Context

This PR integrates the grading system with the participant scores pages for courses and exams

### Description

I've added an average grade and average rated grade on top of the participant scores pages.
In addition, the average scores table now also has an average grade and average rated grade columns.

### Steps for Testing

1. Log in to Artemis
2. Navigate to Course Administration -> Grading Key
3. Create a Grading Key
4. Navigate to Participant scores and verify that (the course should have at least one exercise submission)
a) an average grade and an average rated grade are displayed at the top of the page
b) in the average scores table to last two columns display an average grade and an average rated grade
5. Create a Grading Key for an exam
6. Go through step 4 again for the exam (the exam should have at least one submission)

### Test Coverage

unchanged (100% for both participant scores components)

### Screenshots

Screenshot 2021-07-09 143600

### CODE REVIEW

**Review by willergeorg — CHANGES_REQUESTED** [review]
Tested on TS1. Might not be associated with this PR, but somehow the grading key is not stored after generation (see Screencast below), therefore of course the average grades cannot be displayed as well. I faced this problem for both the course and an exam.

https://user-images.githubusercontent.com/80622272/125158399-30ba4580-e171-11eb-9d69-1c57b4bf255a.mp4

**Comment by ivohashamov** [comment]
> Tested on TS1. Might not be associated with this PR, but somehow the grading key is not stored after generation (see Screencast below), therefore of course the average grades cannot be displayed as well. I faced this problem for both the course and an exam.
> 
>  2021-07-10.11-16-51.mp4

That's odd, I would have to check that out on a test server. I can't reproduce the issue locally. I would have to check it out on a test server. Was the save button constantly disabled and did you test only in your course?

**Review by NicolasUllmann — APPROVED** [review]
Tested on TS1. Everything worked for me, wasn't able to reproduce the error mentioned above.

**Review by willergeorg — DISMISSED** [review]
Tested again on TS1. This time works perfectly fine for me.

**Review by niclasschuemann — CHANGES_REQUESTED** [review]
It seems like there are still translation issues: 
translationIssue
Also, currently, there are no german translations for both labels (German translation is still the English text)

I don't think this is in scope for this PR, but I want to mention it anyway:
The **Participant Scores** Button is present in the course management detail view of a course, but not in the course-management view that lists all courses. Also, while testing I was not quite sure how the average grade is calculated:
gradeweird
I would expect something between 4.7 and 5.0, is it rounded to the next possible grade?

File: src/main/webapp/i18n/de/participantScores.json [hunk 1]
```diff
@@ -30,6 +30,10 @@
             "searchForStudentsOrTeam": "Suche nach Studierende:r oder Team",
             "averageScore": "Durchschnittliches Ergebnis: {{avg}}%",
             "averageRatedScore": "Durchschnittliches bewertetes Ergebnis: {{avg}}%",
+            "averageGrade": "Average Grade",
+            "averageRatedGrade": "Average Rated Grade",
+            "averageBonus": "Average Bonus",
+            "averageRatedBonus": "Average Rated Bonus",
```
Comment (niclasschuemann) [inline src/main/webapp/i18n/de/participantScores.json:36]:
```suggestion
            "averageGrade": "Notendurchschnitt",
            "averageRatedGrade": "Gewerteter Notendurchschnitt",
            "averageBonus": "Durchschnittlicher Bonus",
            "averageRatedBonus": "Gewerteter durchschnittlicher Bonus",
```

**Review by niclasschuemann — CHANGES_REQUESTED** [review]

**Comment by ivohashamov** [comment]
> It seems like there are still translation issues:
> translationIssue
> Also, currently, there are no german translations for both labels (German translation is still the English text)
> 
> I don't think this is in scope for this PR, but I want to mention it anyway:
> The **Participant Scores** Button is present in the course management detail view of a course, but not in the course-management view that lists all courses. Also, while testing I was not quite sure how the average grade is calculated:
> gradeweird
> I would expect something between 4.7 and 5.0, is it rounded to the next possible grade?

@niclasschuemann The translation issue is what Lorena mentioned a few weeks ago in the team meeting. You have to refresh your browser cache and it should work.

Regarding your other comments, I think the Participant Scores button is not on the "overview" pages, but only on the detail ones because it's mainly used for debugging and is not as important.

The average grade is calculated entirely based on the average score -> whatever grade that average score would map to, is the average grade. This is similar to the calculation on the scores pages as well. The reason for that is simple, grades don't necessarily have to be a numeric value -> they could be A,B,C,D,E,F, if that is the type of grading key set. Such grades can't reasonably be averaged. In general, the grading key is discrete in terms of its values, so this gets propagated to the average grade as well.

**Review by canyelten — APPROVED** [review]
Tested on TS1, works perfectly fine.

**Review by ArtemTsurikov — APPROVED** [review]
Works fine on TS1

**Review by katjanakosic — APPROVED** [review]
Works on ts2

**Review by sinangunduz — APPROVED** [review]
tested on ts2, works as expected

**Review by argertboja — APPROVED** [review]
Works fine on TS1!

**Review by imstreetspirit — APPROVED** [review]
tested on ts2, works fine. 

guess, you refactored the code which comes from 404 not found error bugfixing. I also tested as a student. The error is not shown in the console at the end of the exam. Maybe you can also add this into testing steps.

I have one question regarding this new information in the participant scores. So, if do not have a grading key, I can not see the  `average grade` and `average rated grade`. How can I know that to see this results, I should have a grading key as a new user? hope, I expressed myself correctly.

**Pull request merged**